media/open_zim.ksy

```diff
@@ -0,0 +1,356 @@
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::lzma2
+          - id: blobs_ptrs_uncompressed
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::none or info.compression == compression::default
+        instances:
+          first_blob_offset:
+            value: ptr + 1 # info_t.size
+          blobs: 
+            type: blob(_index)
+            repeat: expr
+            repeat-expr: blob_count
+          first_blob_ptr_compressed:
+            pos: first_blob_offset
+            process: kaitai.compress.lzma(2)
+            size: 8 + _root.minimal_xz_lzma_size # max(8, 4) + minimal_xz_lzma_size
+            type: blob_ptr_u
+            if: info.compression == compression::lzma2
+          first_blob_ptr_uncompressed:
+            pos: first_blob_offset
+            type: blob_ptr_u
+            if: info.compression == compression::none or info.compression == compression::default
+          first_blob_ptr_u_v:
+            value: (info.compression == compression::lzma2?first_blob_ptr_compressed:first_blob_ptr_uncompressed).as<blob_ptr_u>.value
+          blob_count_plus_one:
+            value: first_blob_ptr_u_v / info.ptr_size
+          blob_count:
+            value: blob_count_plus_one - 1
+          blobs_ptrs:
+            value: (info.compression == compression::lzma2?blobs_ptrs_compressed:blobs_ptrs_uncompressed).ptrs
+        types:
+          blobs_ptrs:
+            params:
+              - id: count
+                type: u8
+            seq:
+              - id: ptrs
+                type: blob_ptr_u
+                repeat: eos
+            instances:
+              info:
+                value: _parent.info
+          blob:
+            params:
+              - id: idx
+                type: u8
+            instances:
+              data:
+                pos: _parent.ptr + _parent.blobs_ptrs[idx].as<blob_ptr_u>.value
+                size: _parent.blobs_ptrs[idx+1].as<blob_ptr_u>.value - _parent.blobs_ptrs[idx].as<blob_ptr_u>.value
+          info_t:
+            #seq:
+            #  - id: compression
+            #    type: b4
+            #    enum: compression
+            #  - id: ptr_is_u8
+            #    -orig-id: extended
+            #    type: b1
+            #  - id: reserved
+            #    type: b3
+            seq:
+              - id: reserved
+                type: b3
+              - id: ptr_is_u8
+                -orig-id: extended
+                type: b1
+              - id: compression
+                type: b4
+                enum: compression
+            instances:
+              ptr_size:
+                value: (ptr_is_u8?8:4)
+          blob_ptr_u:
+            seq:
+              - id: value
+                type:
+                  switch-on: _parent.as<cluster>.info.ptr_is_u8 # bug in KSC, as is mandatory
+                  cases:
+                    true: u8
+                    false: u4
```

Comment:
What is the bug in KSC? Do you mean that the `value` field here is marked as nullable _(in statically-typed languages like Java, it will have boxed type `Long` instead of primitive `long`)_, despite that _(we humans know)_ the values `true` and `false` cover the full range of possible values of the `ptr_is_u8` (i.e. a `b1` field)?

If so, this has been discussed in https://github.com/kaitai-io/kaitai_struct_formats/pull/303#discussion_r460402466, and the solution is this:

```suggestion
                  switch-on: _parent.as<cluster>.info.ptr_is_u8 # bug in KSC, as is mandatory
                  cases:
                    true: u8
                    _: u4
```

Reply:
>What is the bug in KSC?

If we remove `.as<cluster>`, we get `don't know how to call method 'info' of object type 'CalcKaitaiStructType'`.

About true/false: fixed.

Reply:
> If we remove `.as<cluster>`, we get `don't know how to call method 'info' of object type 'CalcKaitaiStructType'`.

This is not a bug, it is working as intended. See https://doc.kaitai.io/user_guide.html#_enforcing_parent_type - type `blob_ptr_u` may have one of multiple parents (`/types/cluster_ptr/types/cluster` and <code>/types/cluster_ptr/types/cluster/types/**blobs_ptrs**</code>), so they only have the `KaitaiStruct` interface in common (<code>media/open_zim.ksy:{[271-272](https://github.com/kaitai-io/kaitai_struct_formats/blob/208439638843d9ca4e3f656e0964deb173259079/media/open_zim.ksy#L271-L272),[295](https://github.com/kaitai-io/kaitai_struct_formats/blob/208439638843d9ca4e3f656e0964deb173259079/media/open_zim.ksy#L295),[316-319](https://github.com/kaitai-io/kaitai_struct_formats/blob/208439638843d9ca4e3f656e0964deb173259079/media/open_zim.ksy#L316-L319),[359-370](https://github.com/kaitai-io/kaitai_struct_formats/blob/208439638843d9ca4e3f656e0964deb173259079/media/open_zim.ksy#L359-L370)}</code>):

```ksy
    types:
      cluster:
        # ...
        instances:
          # ...
          first_blob_ptr_uncompressed:
            pos: first_blob_offset
            type: blob_ptr_u
            if: info.compression == compression::none or info.compression == compression::default
          # ...
        types:
          blobs_ptrs:
            params:
              - id: count
                type: u8
            seq:
              - id: ptrs
                type: blob_ptr_u
                repeat: eos
```

However, we may definitely improve the error message and make it more useful - in this case, list the types that use `blob_ptr_u` and explain that since it it does not have only one possible parent type, specific attributes of `_parent` are not accessible.